### DISCUSSION TITLE
How Connect MongoDB Atlas to ClickHouse using MongoDB Engine?

### DISCUSSION THREAD

**Discussion by jmtt89** [ClickHouse/ClickHouse#54161] [Q&A]
Hi, i don't find a clear documentation about how connect MongoDB with ClickHouse  

I trying to create a table on ClickHouse that use an Engine to my MongoDB collection.

On Mongo Atlas in find find the URI connection clicking on "connect" button and got something like 

`mongodb+srv://<username>:<password>@my-host.mongodb.net/test?retryWrites=true&w=majority`

i create a table following the Docs and works like a charm. 

```SQL
CREATE TABLE IF NOT EXISTS coll_test
(
    _id UUID,
    name String,
    ...
) ENGINE = MongoDB('my-host.mongodb.net', 'test', 'coll_test', '<username>', '<password>', 'connectTimeoutMS=10000&ssl=true&authSource=admin' );
````

but when i execute some query like

```SQL
SELECT COUNT() FROM coll_test;
```

i receive an error

```
SQL Error [1000] [08000]: Poco::Exception. Code: 1000, e.code() = 0, No address found: my-host.mongodb.net (version 23.7.5.30 (official build))
```

**Comment by jmtt89** [comment]
btw, i know i can connect directly to my primary, using host in direct format `shard-00-00.my-host.mongodb.net:27017` but it's impossible be changing the host every time the primary change....

**Reply by bryzgaloff** [reply]
Are you actually able to connect when using primary's hostname? I get `SSL Exception: error:10000438:SSL routines:OPENSSL_internal:TLSV1_ALERT_INTERNAL_ERROR` on versions `23.10.1.1976` and `23.4.2.11` even though I have enabled access from anywhere (added `0.0.0.0/0` to an access list) in Mongo Atlas. Does it work for you? Or have you encountered `TLSV1_ALERT_INTERNAL_ERROR` issue too and managed to resolve it?

**Reply by jmtt89** [reply]
I did not encounter this error when using the format  shard-00-00.my-host.mongodb.net:27017 , I was able to connect and make queries and load data without problem. 

but for different reasons such as slowness and how complicated it would be to maintain, I decided (at least for now) not to continue with researching...

**Comment by tsolodov** [comment]
mongodb atlas uses SRV records.

you can get A record(or IPs) using `dig -t SRV hostname` and use IPs for connections. 

example:
```
dig -t  SRV serverlessinstance0.xxxxx.mongodb.net

; <<>> DiG 9.10.6 <<>> -t SRV serverlessinstance0.xxxxx.mongodb.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 32670
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4095
;; QUESTION SECTION:
;serverlessinstance0.hlzjbod.mongodb.net. IN SRV

;; ANSWER SECTION:
serverlessinstance0.xxxxxx.mongodb.net. 60 IN SRV 0 0 27017 ac-jumhjgk-lb.xxxxx.mongodb.net.

```

port is open:
```
nc -vz ac-jumhjgk-lb.xxxxx.mongodb.net 27017
Connection to ac-jumhjgk-lb.xxxxx.mongodb.net. port 27017 [tcp/*] succeeded!
```

I am not sure how reliable this approach though

**Comment by wittyphantom333** [comment]
Not sure if related, but I was having similar issues where I could connect, but not return data reliably.  I added this to the end of the ENGINE connection string when creating the table in Clickhouse 

```
authMechanism=DEFAULT&connectTimeoutMS=100000&authSource=admin
```

Now if I can just figure out how to get BinData() from Mongo into a format that doesn't kill my ClickHouse queries I would be golden!

**Comment by Apidcloud** [comment]
What if we need to provide a certificate? It doesn't seem like tlsCAFile is supported

**Reply by tsolodov** [reply]
it looks like it's not possible to use certs for auth

**Reply by Apidcloud** [reply]
It actually is as we found out during the weekend thanks to Robert Hodges (Altinity). By installing the certificate on the operating system (CA trust store), it will be picked up and used when connecting to the mongodb. They are now working on updating the documentation to show this step by step

**Reply by knoguchi** [reply]
Here is what I did with `tlsCAFile` for MongoDB Atlas.  I'm using Ubuntu but the idea should be same for other OS.

```
# Install CA file, in my case it's named my.crt
cp my.crt /usr/local/share/ca-certificates/
update-ca-certificates --verbose
# Add CA file path in /etc/default/clickhouse
echo "SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt" >>/etc/default/clickhouse
# restart ClickHouse
```

Run clickhouse-client, and create table.  Option string may need adjustment but the point is that I set ssl=true only for the TLS connection.
```
CREATE TABLE mongo_example (
_id String
) ENGINE=MongoDB(
    'example.com:27017',  -- Host and port
    'mydb',                               -- Database name
    'mycollection',                           -- Collection name
    'myuser',
    'mypassword',
    'replicaSet=rs0&connectTimeoutMS=10000&authSource=admin&ssl=true'
);
```
